GitHub Server action, to run DA on GitHub

BryceStevenWilley opened the pull request:
~Edit: Blocked by https://github.com/SuffolkLITLab/ALKiln/issues/760.~

In this PR, I have:

* [x] Added tests for any new features or bug fixes (if relevant)
* [x] Added my changes to the CHANGELOG.md at the top, under the "Unreleased" section
* [x] Ensured issues that this PR closes will be [automatically closed](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)

### Reason for this PR

Get things running without external issues.

Closes #756


TODO:
* [x] CHANGELOG
* [x] while loop when the server is starting (no 8 minute sleep)
* [x] research how to call ALKiln action from the same branch.

BryceStevenWilley committed:
Be able to setup a from-scratch server

Starting to look into starting Docassemble servers in Github actions, we needed to:
* log-in to the server
* create an API key, and save it
* use said API key to install the package, so we have all of the dependencies
  This currently can install things, but can't properly keep track of the task_id
  for some reason? Need to look into it

Also started the full bash script (which would become the Github actions script)
for the process.

BryceStevenWilley committed:
Temp trying some stuff

BryceStevenWilley committed:
Separate package restart and playground restart, and work on installed packages

BryceStevenWilley committed:
Run assembly Line tests, not toolbox

BryceStevenWilley committed:
Don't takedown from the project, nothing is there

BryceStevenWilley committed:
Add github_server action

In progress! Much of the existing work came from https://github.com/BryceStevenWilley/SmolKilnPOC/blob/main/.github/workflows/smol.yml

plocket commented on lib/utils/session_vars.js:
```diff
@@ -197,6 +198,36 @@ session_vars.save_artifacts_path_name = function ( path_name ) {
 };  // Ends save_artifacts_path_name()
 
+
+session_vars.save_install_type = function (install_type) {
```
From what I can see, this can possibly be taken care of with the `_ORIGIN` env var. Is there a reason it needs to be saved into the runtime config instead that I'm not seeing?

plocket replied:
I see now that `install_type` is different in definition than `_ORIGIN` conceptually. `install_type` is if it was installed in the playground vs. on the server. That is a more clear description. I'm wondering if it's still enough to just infer it from `_ORIGIN` or if we really want to introduce a new action step-crossing config variable. I'll restore the previous code, but I think it's worth discussing.

BryceStevenWilley replied:
1) I think of `_ORIGIN` and `install_type` as separate things: `_ORIGIN` should only deal with where Kiln JS is running, and `install_type` is independent of that. I can easily see GitHub actions or local runs using either installed interviews or playground interviews. I agree it's less likely to run KilnInThePlayground on an installed interview, but...

2) I still hold that `_ORIGIN` is a temporary solution, and the problem it solves is better and more flexibly solved by command args for parameters, so I've been avoiding building more functionality on top of it.

plocket replied:
Deep dive discussion: Since the two vars could be used in combination someday, we'll keep them separate for now. We'll also add the `install_type` interview url to the arguments that will be passed in through the command line when we do the same for the file names (avoiding `_ORIGIN`, though we do have to check if that env var is used for something else too), so for now we'll keep things as they are since they'll be changed anyway.

plocket replied:
Conclusion: We used `install_method` and kept `_ORIGIN` separate.

plocket committed:
Restore action.yml to v5

plocket committed:
Restore error for missing env vars

plocket committed:
Replace runtime_config.json var with _ORIGIN env var. Not sure

if the var really does need to be saved in the config for storage
in between runs or in between setup, test, and takedown.

plocket commented on github_server/action.yml:
```diff
@@ -0,0 +1,53 @@
+  MAX_SECONDS_FOR_SETUP:
+    description: "Amount of time to give the Project to upload your package's GitHub code."
+    required: false
+    default: '120'
```
Note to self: Remove this default, like we did for the v5 normal action.

plocket committed:
Revert "Replace runtime_config.json var with _ORIGIN env var. Not sure"

This reverts commit 0507540f51d1c662c6f3caa38254844f8814ba5c.

plocket commented on github_server/action.yml:
```diff
@@ -0,0 +1,53 @@
+description: "Automatically test docassemble interviews on GitHub's servers, in an isolated environment. Avoids docassemble reload errors and other external factors"
+
+
+inputs:
```
TODO: Do the regular ALKiln `inputs` of the package authors need to be handled in any special way to pass them on to the default ALKiln action? [This should become obvious if we test it with a package that needs some inputs.]

BryceStevenWilley replied:
Secrets should _not_ be affected, as they're all read from env vars, but we do need to manually pass these args to the `uses: SuffolkLITLab/ALKiln` step below

plocket replied:
Can confirm. [For future discussion:] This duplicates the work we do in ALKiln. I wonder if there's another way to do this. Also, I'm concerned about the custom env vars that users are setting - will they come through. I suspect they'll be fine, but we'll be able to test that later. Given that, would it be better to have authors set the ALKiln as env vars instead of passing them as inputs. It would be a breaking change, but it would save a lot of duplication and maintenance cost.

plocket replied:
Possible argument for keeping inputs: It makes those vars distinct from other vars - they clearly belong to ALKiln and we are able to make some required. Maybe they're also (only a little) easier to look up in the documentation. I'm not as sure about the ones that aren't required.

plocket replied:
Conclusion: (not about inputs in general) We made a separate workflow script that authors can make to use the server action. In their workflow, the set up the server with one action and then run ALKiln with the other action, with special arguments.

plocket commented:
Note: need to know what different errors look like in this process and document them (like missing env vars, missing config stuff, missing package dependencies, etc.)

BryceStevenWilley commented on github_server/action.yml:
```diff
@@ -0,0 +1,53 @@
+          echo "$CONFIG_CONTENTS" > /tmp/config/myconfig.yml 
+          docker pull brycewilley/smoldocassemble
+          docker run --env DA_ADMIN_EMAIL="admin@example.com" --env DA_ADMIN_PASSWORD="password" --env DA_ADMIN_API_KEY="${{ env.DA_ADMIN_API_KEY }}" --env DA_CONFIG=/tmp/config/myconfig.yml --volume /tmp/config:/tmp/config --cap-add SYS_PTRACE --memory="4gb" -d -p 80:80 jhpyle/docassemble
+          # TODO before merge: do a while loop here until `localhost` shows "It's ready"
+          sleep 480
+          curl localhost
```
A thing that should change before this gets merged; we shouldn't need a huge sleep. It should do some loop with a few seconds of sleep (3-5 probably), and then `curl localhost` to see if we get served the "It's ready" page, or the "docassemble is loading...".

BryceStevenWilley commented on github_server/action.yml:
```diff
@@ -0,0 +1,53 @@
+          DOCASSEMBLE_DEVELOPER_API_KEY: "${{ env.DA_ADMIN_API_KEY }}"
+      - name: Docker logs
+        # TODO: create an arifact for this instead
+        run: |
+          docker logs $(docker ps --filter "ancestor=jhpyle/docassemble" --format {{.ID}})
+        shell: bash
```
Another refactor: the logs could have a few issues:

* this won't print if the ALKiln tests or the previous step fails for any reason. it should print in those cases too
* there is the chance that it prints secrets from the config file. The way that we set the config file is through the entire thing, which means I think that GitHub will only censor things if it prints the config in it's entirety, and won't censor specific lines. Some ways around that include:
  * making the config file a mako template or something similar, and replacing the actual secret values with specific github secrets. That would mean that github would still censor the actual secret parts in case they're printed in the log.
  * seeing if GitHub lets us restrict who can download artifact files, and making this a downloadable file instead of printing straight to the GH log?

plocket replied:
I've at least confirmed that a user who isn't logged in can't download the artifacts or see the details of the console log. I assume you mean restricting on a more fine-grained basis. E.g. only allow admins to see those.

plocket replied:
This is what I'm told about config logs:

> The configuration isn't echoed to output, but error messages might contain references to configuration values.

plocket replied:
Another alternative (not that I'm a huge fan) is to encrypt the output file. It's a lot more work for the user, but so far I haven't found a way to restrict permissions.

As for Mako + secrets, just to state it explicitly, that's up to the user to do in their own repository. I'm not sure what those files would look like and how you'd run those scripts. That, too, would be a lot of work for the user. Maybe more work than an encrypted file.

BryceStevenWilley commented on lib/docassemble/get_base_interview_url.js:
```diff
@@ -11,9 +11,18 @@ let get_base_interview_url = async function () {
-  return `${ server_url }/interview?new_session=1&i=docassemble.playground${ dev_id }${ project_name }%3A`;
+  console.log(session_vars.get_install_type());
+  if (session_vars.get_install_type() === "server") {
+    let x = `${ server_url }/interview?new_session=1&i=${session_vars.get_package_name().replace("-", ".")}%3A`;
+    console.log(`base_interview_url: ${x}`)
+    return x;
+  } else {
+    let dev_id = await docassemble_api_interface.get_dev_id();
+    let project_name = session_vars.get_project_name();
+    let x = `${ server_url }/interview?new_session=1&i=docassemble.playground${ dev_id }${ project_name }%3A`;
```
Discussed on deep dive, but another refactor I would consider if there should be a different way of passing along this URL. Not sure the best way though.

plocket replied:
Thoughts: We're always going to be the ones deriving the URL. I think we were just describing separating the concerns. We could make a script that just determines the url of an interview installed on the server and a script that determines the url of a playground interview. Each action could use the appropriate one and output the result to the action, which would then use it in the command line arguments.

plocket replied:
We'll have to have a real-time discussion to refresh my memory on this. Right now my confused brain is wondering why we need to create the whole url ourselves. For example, couldn't we send the base url of the interview through the command line [- for the server `serverurl/interview?i=`, playground same plus `docasemble.playground`]. Maybe that's still us - ALKilnInThePlayground is us - but there's still a distinction, right? Allow that repo to send the base url through the command line? Is that the flexibility we're looking for? So that we, as developers [running the command manually], can pass in the base url of the interview through the command line?

Maybe we should write some user stories to help clarify the exact paths we're thinking of.

plocket replied:
Conclusion: There were only 2 url types - server install and playground install, so we'll handle it by using an npm script to install a runtime config var we can use as a flag.

BryceStevenWilley commented on lib/docassemble/startserver.js:
```diff
@@ -0,0 +1,69 @@
+  await page.goto(`${ session_vars.get_da_server_url() }/user/sign-out`, {waitUntil: `domcontentloaded`});  
+  console.log(`signed out`);
+  await browser.close();
+  */
```
IDK if any of the above is useful, but I did have a version that could make an API key for the user, but it still needs a user name and password.

plocket replied:
If we do want to create unique API keys and we do want to keep them secret, we'd need to work with the GitHub API to create those secrets.

Stream of consciousness: Not sure why we need a custom API key - the server gets killed after the tests. Are we concerned people will interfere with the process in the middle of the action? I think only people with permissions could use something like tmate with a test run. It seems like that could be a security measure the authors have to take - only allow trusted folks to be collaborators. Maybe that's too restrictive? If so, does GitHub have a way to have finer-grained permissions - allow only admins to mess with runningtests? Even that might be too restrictive since collaborators do need to look at tests, though I guess they don't need to look at running tests...

BryceStevenWilley replied:
Whatever API key we make, we shouldn't keep around, so don't need to worry about that.

And I only brought this up in case it might be useful for other Kiln features; it would not be useful for running docassemble on docker on GitHub.

plocket replied:
It's true it's another way to make an api key. I wish there were a place to store code like this without keeping it in the repo itself. I'll never find it in here. Maybe we can make issues for this kind of stuff with a special label or something.

plocket replied:
Conclusion: Made and closed https://github.com/SuffolkLITLab/ALKiln/issues/815

BryceStevenWilley commented on tmp.bash:
can remove this file, I don't remember what it existed for

BryceStevenWilley commented on lib/docassemble/startserver.js:
worth thinking of a different name for this file; startserver isn't great, likely `setup-install` or something closer to the other setup

plocket replied:
Maybe `server_install`.

plocket commented:
I'll put the notes of my action experiments on https://github.com/SuffolkLITLab/ALKiln/tree/start_server_workflow in here so everything's in the same place.

The biggest issue I see as I work through this is that we run the github_server setup to get the interview on the server, but then we also run the regular setup and takedown, which will install the repo on the Playground. Do we want to double up that way? Is there a point to doing both? Where do we want to control whether those happen or not?

Testing bugs so far: In an [08/14 test run](https://github.com/SuffolkLITLab/ALKiln/actions/runs/5863230862/job/15896328024):
- A wait of 700 still [didn't give the server enough time to start](https://github.com/SuffolkLITLab/ALKiln/actions/runs/5863230862/job/15896328024#step:3:164) (the curl happens after the sleep, so it should be showing that the server has already started, which it isn't). Bryce mentioned that we may not be able to look at what's going wrong very easily, so we'll have to figure out the next steps.
- As suspected, `SERVER_URL` and friends were not defined in the ALKiln action itself: https://github.com/SuffolkLITLab/ALKiln/actions/runs/5863230862/job/15896328024#step:3:198. (at alkiln-startserver, where `session_vars` is first used).
- `REPO_URL` also wasn't defined, which makes sense, but means we need to duplicate some of the work we're doing in the regular `action.yml`. It is needed in `da_i.install_on_server`. I'm not loving the amount of duplication we're having to do.

Thoughts:

Maybe we need to somehow break out into further actions to avoid duplication - an action for accepting input and creating env vars that happens before the other actions. It, or something else, would then need to manage calling the right follow-up actions. If we do that, the github server action can avoid the regular `action.yml` setup and takedown by just calling `test` ourselves.

This seems complicated and would require conditional logic with another env var or input, which we were trying to avoid by splitting these actions into separate scripts (as we discussed). We should discuss which is worse - the duplication or the additional input/env var. (I'm reluctant to use another input as we're running into duplication problems with our current ones.)

BryceStevenWilley commented:
Seeing the state of things and the fact that we do still need to export / duplicate the env vars in both `actions.yml`s for things to be fully separate , I think the extra input to `ALKiln/action.yml` isn't that bad of an option; it can be optional, so it won't change any one's existing usage, and can default to the normal usage.

> we're running into duplication problems with our current ones

If you're talking about v5 vs v4 duplication issues, this is v5 only.

BryceStevenWilley commented on github_server/action.yml:
```diff
@@ -0,0 +1,53 @@
+        run: |
+          mkdir /tmp/config
+          echo "$CONFIG_CONTENTS" > /tmp/config/myconfig.yml 
+          docker pull brycewilley/smoldocassemble
```
```suggestion
          docker pull jhpyle/docassemble
```

This will avoid an extra 1-2 minutes of downloading a docker image that we aren't using anymore (though it's worth considering trying that again if it keeps taking that long for everything to start up).

plocket committed:
Implement GitHub server workflow and action (#753)

Implement GitHub server workflow and action

* Start new workflow file github_server.yml

Create github_server.yml

🤞

Use branch with the new action in the new workflow

Attempt to install package and test only two tests

Double the waiting time

Switch the action branch to the current branch

Publish and use alkiln-startserver bin

Install ALKiln with npm before using its darn script

* Use full jhpyle da image.

Co-authored-by: BryceStevenWilley <Bryce.Steven.Willey@gmail.com>

* Try to always print the final log, remove MAX default

* Use custom config, output values, wait for docker to start

Add tmate

Separate wait into next step

Remove conditional tmate

Remove sudo

Allow action to continue

Lengthen sleep time to give us time with detached tmate

Use input correctly (I hope...)

Avoid running ALKiln in the composite action and...

- Only start the server in the github_server composite action.
- Create outputs in that action. Print them in the workflow that
uses that composite action.
- Use a while loop to detect whether docker is done starting.
- Make a placeholder for a max docker start time input.

Remove `needs`. LLM's aren't perfect.

* Use INSTALL_METHOD input to control action.yml behavior

Decrease initial server wait time, try to run tests

Docker start time was much shorter than expected, so decreasing
the amount of hard-coded time to wait before starting to test
whether the server is loaded or not.

Also, used a different variable for stopping the while loop so
that race conditions don't prevent the success message from
getting logged.

Try to use var differently

Try different syntax for var

Try keeping env var. Can't use inputs just anywhere.

Use the correct alkiln version for the command

* Use relative paths for `uses:` to avoid editing val on every branch change

Use branch ref in `uses`, hard code tags for now, tweak var names,

clean up a bit

Try different var syntax

Try to use the full repo/branch ref as an env var

Try a full path instead

Go out two folders

Try path code given by llm phind

Try relative path again

Remove the action.yml filename, just reference the dir

Try to use var again

Use relative path again, again going out 2 dirs

Use tmate

Use tmate better

Just run tmate

Remove the reference to the file itself

* Use correct sign in step info for actual tests

* Use `dainstall` and health check, save docker artifact

Add tempt tmate to ALKiln test run to find interview on server

Try different base url ending in data/questions/

Bump version for new base interview url

Refine interview base url

Log full interview url

Our package name is different than our project name. Problem?

Bump version for package url test

Explore folders with tmate

Looking for something like /usr/share/docassemble/local3.10/lib/python3.10/site-packages/docassemble/ALKiln/data/questions

Match "ALKiln" folder name to package name as pip hasn't been able

to install the package and we're theorizing that that's the issue.

Change ALKilnTests to ALKiln, remove tmp dir, clean

Add tmate step to experiment with dainstall

Try to install using dainstall

Isolate repo name

Try new string substitution syntax

Fix typo

Remove debug log because its error is not worth solving

Run tests again since installation seems to be working

Actually activate tests

Troubleshoot syntax that was working before

Tweak indentation

Allow server install address without js installing on server

Bump feature version

Switch back to hard-coded API key

Set `result` in gh_server action

Add AL requirement

Add dangerous debug for debugging server action, clean up

Test the failing test

Log dangerous vars

Bump for log

Try a different password

Try to use output email and password as env vars

Log container name

Create logs artifact, attempt 1

Typo

Use absolute path for artifact

Change startserver -> server_install file/function name

Bump for file/function name change

Start exploring url/healt_check output

Try to disambiguate log of health check

Try to get just the status code

Tangent to try installing the package without cloning

Add shell. There must be something global.

Move reference to dir path for dainstall

Add in a test that actually uses the package

Use health_check response to check docker finished starting

Conform to github action output guidelines

Test docker install input

Restoring defaults as the input test was successful

Test all our tests

* Make missing file error test more broad

Make missing file error test more broad

Bump for missing file test and console log

* Remove `install_on_server` API request

* Use dainstall to install repo on server and...

use folder name to create interview address as the repo folder can have
different name from repo name. Leave interviews installed on remote
(non-github folders) and their installation process as it is.

This is a combination of 11 commits.

Change docassemble package folder back to ALKilnTests

Add tmate to experiment with #egg

Try to use inner folder name to get server location of interview

Log more info about the chosen folders. Also try to use fewer tests

Fail tag definition faster

Bump dev to test with right ALKiln version. Pass tags as input.

Add new ALKiln version

Fix typo in function name

Fix typo, use @server to install instead of @v#

Fix requested version number typo

Parse folder names differently

* Switch install_type -> install_method, abstract config save/get

* Run all ALKiln tests

* Apply PR suggestions

- Simplify getting docker logs now using our container name, though maybe this needs to use the var name
- `exit 1` on docker error
- Typo

Co-authored-by: Bryce Willey <Bryce.Steven.Willey@gmail.com>

* Remove the var `ready`, comments and cleanup

* Name the docker container (not getting output)

Also temp reduce to one test for testing output

* Handle no config value. Edit comments and clean.

* Re-activate all github server tests, tweak logs and comments

- Tweak logs and add comments
- Re-activate all github server tests
- More small clean up

* Tweak folder name, improve console logs, date/time foldername,

and clean up

* Remove dangerous debug

* Address docker log filename bash error with quotes and env var

Address docker log filename bash error with quotes and env var

Address docker log filename bash error with double quotes

Address filename error (date?) with env var `TZ` instead of -u arg

Restore one of the changes. Perhaps the bash is different in each

place? Though they both use bash.

Try double quotes

Use env var for filename to allow download to find the right file

Use env var for filename to allow download to find the right file

---------

Co-authored-by: BryceStevenWilley <Bryce.Steven.Willey@gmail.com>

plocket commented:
Docassemble core is giving the wrong error. This is blocked till the next release when that should be fixed.

plocket committed:
Update CHANGELOG

plocket reviewed (approved):
I'm a bit too close to it, but from what I can tell we're good to put it out there.
